### ISSUE TITLE
Freeze on multiple GPUs with CustomNonbondedForce with long range correction

### ISSUE THREAD

**Issue by SHervoe-Hansen** [openmm/openmm#3661]
So I have experimented with the .xml forcefield format for openMM and came across a huge slow-down in performance upon inclusions of `NBFixPair`.

An example that has been able to reproduce the error on two different machines so far is the following:
```python
import openmm as mm
from openmm import app
from openmm import unit as u

print('Loading initial configuration and toplogy')
pdb = app.PDBFile('NH4SO4_0.50_M.pdb')
forcefield = app.ForceField('tip3p_standard.xml', 'NH4_mod.xml', 'SO4_mod.xml', 'Exceptions.xml')

modeller = app.Modeller(pdb.topology, pdb.positions)
modeller.addHydrogens(forcefield, pH=7.0)
modeller.addSolvent(forcefield, model='tip3p', numAdded=29000, neutralize=False)
system = forcefield.createSystem(modeller.topology, nonbondedMethod=app.PME, ewaldErrorTolerance=0.0005, switchDistance=1*u.nanometer, nonbondedCutoff=1.2*u.nanometers, constraints=app.HBonds, rigidWater=True, hydrogenMass=1.25*u.dalton)

total_mass = u.sum([system.getParticleMass(i) for i in range(system.getNumParticles())])
integrator = mm.LangevinMiddleIntegrator(298.15*u.kelvin, 1.0/u.picoseconds, 4.0*u.femtoseconds)
system.addForce(mm.MonteCarloBarostat(1*u.bar, 298.15*u.kelvin, 25))
platform = mm.Platform.getPlatformByName('CUDA')
properties = {'CudaPrecision': 'mixed', 'CudaDeviceIndex': '0,1'}
sim = app.Simulation(modeller.topology, system, integrator, platform, properties)
sim.context.setPositions(modeller.positions)

print('Calculating energy')
state = sim.context.getState(getEnergy=True, getForces=True) # It will never get past this step!
print('Starting pot energy:', state.getPotentialEnergy().in_units_of(u.kilocalories_per_mole))
```
Changing `properties = {'CudaPrecision': 'mixed', 'CudaDeviceIndex': '0,1'}` to `properties = {'CudaPrecision': 'mixed', 'CudaDeviceIndex': '0'}` or removing `Exceptions.xml` from `app.ForceField` causes the system to yield expected performance.

The content of `Exceptions.xml` is
```xml
<ForceField>
    <LennardJonesForce lj14scale="0.5">
        <!--tip3p water types-->
        <Atom type="tip3p-O" sigma="0.31507524065751241" epsilon="0.635968"/>
        <Atom type="tip3p-H" sigma="1" epsilon="0"/>
        <!-- ammonium sulfate types-->
        <Atom type="ammonium-N" sigma="3.25000e-01" epsilon="7.11280e-01"/>
        <Atom type="ammonium-H" sigma="1.06908e-01" epsilon="6.56888e-02"/>
        <Atom type="sulfate-S" sigma="3.56359e-01" epsilon="1.04600e+00"/>
        <Atom type="sulfate-O" sigma="2.95992e-01" epsilon="8.78640e-01"/>
        <NBFixPair type1="sulfate-O" type2="ammonium-N" sigma="3.50860e-1" epsilon="7.90543e-1"/>
        <NBFixPair type1="sulfate-O" type2="ammonium-H" sigma="2.27638e-1" epsilon="2.40243e-1"/>
    </LennardJonesForce>
</ForceField>
```
Side question: Is it a requirement all types must be specified if one includes `LennardJonesForce` or `NBFixPair`? I get an error if I do not include the water parameters.

Looking forward to hearing from you

**Comment by peastman** [comment]
That sounds like you're getting a hang, not a slowdown.  Can you provide a complete test case I can use to reproduce it?

Using an NBFix shouldn't have any effect on the calculation.  It uses lookup tables for LJ parameters.  The only effect of the pairs is to change the particular values stored in those tables.

> Is it a requirement all types must be specified if one includes LennardJonesForce or NBFixPair?

Yes.  It's a nonbonded force in which every particle interacts with every other.  That means it needs parameters for every particle.

**Comment by SHervoe-Hansen** [comment]
You can find the complete test case I use
[ here (openMM-hang.tar.gz)](https://github.com/openmm/openmm/files/8978526/openMM-hang.tar.gz).

I have been running using openMM version 7.7,  CUDA version 11.6 on Tesla V100s.

**Comment by peastman** [comment]
It runs fine for me:

```
Loading initial configuration and toplogy
Creating OpenMM System
Calculating energy
Starting pot energy: 26031.08315966132 kcal/mol
Minimizing energy
```

There's no visible delay between "Calculating energy" and printing the result.  The only obvious difference is that I'm running on a single GPU (a Titan V), though I don't know any reason that would matter.  Can you try using a single GPU?

**Comment by SHervoe-Hansen** [comment]
>Changing properties = {'CudaPrecision': 'mixed', 'CudaDeviceIndex': '0,1'} to properties = {'CudaPrecision': 'mixed', 'CudaDeviceIndex': '0'} or removing Exceptions.xml from app.ForceField causes the system to yield expected performance.

Running on a single GPU also works fine for me. The problem only occurs when running or two or more.

**Comment by SHervoe-Hansen** [comment]
I don't know if this might got something to do with it, but my forcefields does include both a NonbondedForce and LennardJonesForce section in which the same atoms are specified. I hope this does not mean it applies double LJ forces on the atoms.

**Comment by peastman** [comment]
I don't know any reason that would cause it.  Your forces will be wrong, which will probably cause the simulation to blow up.  But that shouldn't prevent it from just calculating the energy.

To prevent the NonbondedForce from calculating LJ, set `epsilon` to 0 for every particle.

**Comment by peastman** [comment]
I think I understand the problem.  I need to think a bit to figure out how to resolve it.

The particular combination of factors that triggers it is 1) using multiple GPUs, and 2) having a CustomNonbondedForce with long range correction enabled.  When the coefficient for the long range correction needs to be recomputed (or computed for the first time at the start of the simulation), it's done in the background on the context's worker thread.  At the end of the energy computation, `CudaCalcForcesAndEnergyKernel::finishComputation()` invokes a post-computation that calls `flush()` on the worker thread to wait until the computation is done, then returns the result.

But when you're using multiple GPUs, `finishComputation()` is itself run on the worker thread.  So the worker thread blocks waiting for the worker thread to complete all its work...

There's a few approaches we might use for resolving this.  Let me think about it.

**Comment by peastman** [comment]
The fix is in #3668.
